### PR TITLE
Don't capture HttpContext into the log objects

### LINKED ISSUE #10556: HostingRequestStartingLog shouldn't capture HttpRequest for its values
`HostingRequestStartingLog` captures `HttpRequest` for its values and resolves them from that.

If the logging happens lazily/out-of-line with the request (ConsoleLogger for example queues them) then the logging can happen outside of the request and the values be incorrect or reading them cause a race condition.

### LINKED ISSUE #10557: HostingRequestFinishedLog shouldn't capture HttpContext for its values
`HostingRequestFinishedLog` captures `HttpContext` for its values and resolves them from that.

If the logging happens lazily/out-of-line with the request (ConsoleLogger for example queues them) then the logging can happen outside of the request and the values be incorrect or reading them cause a race condition.

### PR DESCRIPTION
**Pull request by benaadams**
Don't capture HttpContext into the log objects from https://github.com/aspnet/AspNetCore/pull/10319

Resolves #10556
Resolves #10557

/cc @davidfowl

### CODE REVIEW

File: src/Hosting/Hosting/src/Internal/HostingRequestStartingLog.cs [hunk 1]
```diff
@@ -51,7 +59,17 @@ internal class HostingRequestStartingLog : IReadOnlyList<KeyValuePair<string, ob
 
         public HostingRequestStartingLog(HttpContext httpContext)
         {
-            _request = httpContext.Request;
+            var request = httpContext.Request;
+
+            Protocol = request.Protocol;
+            Method = request.Method;
+            ContentType = !string.IsNullOrEmpty(request.ContentType) ? Uri.EscapeUriString(request.ContentType) : string.Empty;
```
Comment (davidfowl) [inline src/Hosting/Hosting/src/Internal/HostingRequestStartingLog.cs]: Allocations!

**Comment by davidfowl** [comment]
I think this change is fine but there's no issue with console and other loggers that store messages in a queue for further processing (at least the ones that we ship). They materialize the message before en-queuing.

**Comment by benaadams** [comment]
>  there's no issue with console and other loggers that store messages

If you are happy with the current state, I'll close this as it does make `HostingRequestStartingLog` a bit bloaty

**Comment by neuecc** [comment]
@davidfowl @benaadams 

The logger I'm developing has encountered this issue:
https://github.com/Cysharp/ZLogger

ZLogger adopts an architecture that consistently delays formatting and writes to an `IBufferWriter<byte>` to avoid allocating strings each time and for performance batching.
As a result, this logging, which captures the HttpContext, is causing problems:
https://github.com/Cysharp/ZLogger/issues/165

Looking at the PR, it seems that incorporating this would solve the issue.
Is there still a possibility of reconsidering and merging this?

Also, I think it's bad to capture the HttpContext.
I were aware that the presence of such a capture could cause problems, but I did not believe that there was anything in the standard that would do such a behavior.

**Pull request closed without merging**